### PR TITLE
Shader Reflection implementation V3

### PR DESCRIPTION
**Pull request by jaynus**
This is a continuation of #72 

This pull request includes:
 - [x] Working Spirv Reflection for basic example (see quads example)
 - [x] spirv-reflect -> gfx_hal middleware layer for converting the appropriate descriptor and shader information
- [x] Changes in rendy_shader for the reflection
- [x] All changes are feature-gated with spriv-reflection feature, enabled by default
- [x] Support for push constants
- [x] Test cases for layout merging
- [x] Example documentation
- [x] Name-based attribute reflection
- [x] Simple API for reflection

Moved to a fresh branch and new PR to have a clean slate for review.

### CODE REVIEW

File: shader/src/lib.rs [hunk 1]
```diff
@@ -46,19 +69,380 @@ pub trait Shader {
...
+        self.stage
+    }
+}
+
+/// A `ShaderSet` object represents a merged collection of `ShaderStorage` structures, which reflects merged information for all shaders in the set.
+#[derive(derivative::Derivative, Debug)]
+#[derivative(Default(bound = ""))]
+pub struct ShaderSet<B: Backend> {
+    shaders: HashMap<ShaderStageFlags, ShaderStorage<B>>,
+    set_reflection: SpirvReflection,
```
[the remark this answers is not in the document]
Comment (jaynus) [reply shader/src/lib.rs]: removed.

File: shader/src/lib.rs [hunk 2]
```diff
@@ -46,19 +69,380 @@ pub trait Shader {
...
+    ///
+    /// # Parameters
+    ///
+    /// `factory`   - factory to create shader modules.
+    ///
+    pub fn build<B: Backend>(
+        &self,
+        factory: &rendy_factory::Factory<B>,
+    ) -> Result<ShaderSet<B>, failure::Error> {
+        let mut set = ShaderSet::<B>::default();
```
[the remark this answers is not in the document]
Comment (jaynus) [reply shader/src/lib.rs]: removed.

File: shader/src/lib.rs [hunk 3]
```diff
@@ -46,19 +69,380 @@ pub trait Shader {
...
+            unsafe {
+                storage.compile(factory)?;
+            }
+            Ok(storage)
+        };
+
+        if let Some(shader) = self.vertex.clone() {
+            set.shaders.insert(
+                ShaderStageFlags::VERTEX,
+                create_storage(ShaderStageFlags::VERTEX, shader, factory)?,
```
[the remark this answers is not in the document]
Comment (jaynus) [reply shader/src/lib.rs]: Refactored.

File: shader/src/reflect/mod.rs [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    }
+}
+
+impl<B: Backend> SpirvReflectionGenerator for ShaderSet<B> {
+    #[inline(always)]
+    fn attributes_range<R: RangeBounds<usize>>(
+        &self,
+        range: R,
+    ) -> Result<VertexFormat, failure::Error> {
+        self.set_reflection.attributes_range(range)
```
[the remark this answers is not in the document]
Comment (jaynus) [reply shader/src/reflect/mod.rs]: removed

File: rendy/examples/meshes/main.rs [hunk 5]
```diff
@@ -57,6 +63,22 @@ lazy_static::lazy_static! {
     ).precompile().unwrap();
 }
 
+#[cfg(feature = "spirv-reflection")]
+lazy_static::lazy_static! {
+    static ref SHADERS: rendy::shader::ShaderSetBuilder = rendy::shader::ShaderSetBuilder::default()
+        .with_vertex(&*VERTEX).unwrap()
+        .with_fragment(&*FRAGMENT).unwrap();
+
+    static ref SHADER_REFLECTION: SpirvReflection = SHADERS.reflect().unwrap();
```
Comment (Frizi) [inline rendy/examples/meshes/main.rs]: you can put ` #[cfg(feature = "spirv-reflection")]` directly on those statics now :)

File: rendy/examples/meshes/main.rs [hunk 6]
```diff
@@ -329,13 +355,30 @@ where
             Some(self.sets[index].raw()),
             std::iter::empty(),
         );
+
+        #[cfg(feature = "spirv-reflection")]
```
Comment (Frizi) [inline rendy/examples/meshes/main.rs]: you can now reduce such long feature based repetitions with
```rust
#[cfg(feature = "spirv-reflection")]
let vertex = SHADER_REFLECTION.attributes(&["pos", "color", "norm"]).unwrap();
#[cfg(not(feature = "spirv-reflection"))]
let vertex = PosColorNorm::VERTEX;
...
.bind(&[vertex], &mut encoder)
```

**Comment by jaynus** [comment]
bors retry

**Pull request closed without merging**